Add EastWest Range feature

jonathanbeber opened the pull request:
The current east-west communication feature has some problems as the
lack of support for cluster-only routes and the fact that these internal
routes would still be accessible from outside if changing the Host
header. More details of #1526

This commit adds the concept of EastWest Range feature, what makes
possible to define internal domains and custom predicates to Skipper.
With this information, Skipper's Kubernetes dataclient adds the given
predicates on every route identified as part of one internal domain.
Differently from the `-enable-kubernetes-east-west` and the
`-kubernetes-east-west-domain=.ingress.cluster.local` flags this feature
will not automatically create routes and both features shouldn't be used
in combination.

It requires changing the current way we convert routegroups to eskip
routes, splitting the hostnames in internal and external hosts. All the
hosts identified as internals require a separated route once it will
have their own `ClientIP` predicate.

Also, it adds two new flags to skipper configuration,
`-kubernetes-east-west-range-domains` and
`-kubernetes-east-west-range-predicates`. It also include the basic
documentation regarding their usage.

szuecs commented on config/config_test.go:
```diff
@@ -64,6 +64,8 @@ func Test_NewConfig(t *testing.T) {
 				PrependFilters:                          &defaultFiltersFlags{},
 				SourcePollTimeout:                       3000,
+				KubernetesEastWestRangeString:           commaListFlag(),
+				KubernetesEastWestRange:                 map[string]string{},
```
Should be nil instead, if it doesn’t work, needs fix in code

szuecs commented on dataclients/kubernetes/eastwest.go:
```diff
@@ -43,3 +46,52 @@ func createEastWestRouteRG(name, ns, postfix string, r *eskip.Route) *eskip.Rout
 }
+
+func applyEastWestRange(eWRange map[string]string, host string, routes []*eskip.Route) {
+	for d, n := range eWRange {
```
d is domain, n is cidr?
d seems fine, n is weird.

jonathanbeber replied:
should be network

jonathanbeber replied:
will change.

szuecs commented on dataclients/kubernetes/eastwest.go:
```diff
@@ -43,3 +46,52 @@ func createEastWestRouteRG(name, ns, postfix string, r *eskip.Route) *eskip.Rout
+func applyEastWestRangePredicates(routes []*eskip.Route, n string) {
+	for _, route := range routes {
+		route.Predicates = append(route.Predicates, &eskip.Predicate{
+			Name: "SourceFromLast",
```
That’s the wrong predicate. We only want to allow https://opensource.zalando.com/skipper/reference/predicates/#clientip without xff headers. Within cluster doesn’t pass the cloud loadbalancer

jonathanbeber replied:
good one. I will update the narrative too.

jonathanbeber replied:
for the record, I'm not sure if by 
> without xff headers

you were referring to check whether the header is present, but in any case, that's not necessary since faking the `Remote-Addr` would result in a failed syn-ack.

mikkeloscar replied:
@jonathanbeber I think sandor is referring to the fact that if you have an NLB (not ALB) in front of skipper then you can fake the xff header if `SourceFromLast` is used. This is why we created `ClientIP` which was done quite recently #1614 :)

mikkeloscar replied:
That's also the reason `SourceFromLast` was mentioned in the internal narrative, because we didn't have anything else at the time and didn't yet consider NLB as an option.

jonathanbeber replied:
Thanks, @mikkeloscar, then, indeed, using just `ClientIP` is considered enough, I'm marking it as resolved.

jonathanbeber replied:
reopening it as after some test we found out that just `ClientIP` won't work for us: https://github.com/zalando/skipper/issues/1526#issuecomment-756903746

jonathanbeber replied:
this means we will also have to support multiple CIDRs per domain (or at least two).

szuecs replied:
Multiple cidr makes sense for migrations, too. Think about migration of subnetwork we already did. to increase the network size for clusters.

jonathanbeber replied:
it's done now, it accepts predicates instead of CIDRs, what gives us a lot more flexibility.

szuecs commented on dataclients/kubernetes/eastwest.go:
```diff
@@ -43,3 +46,52 @@ func createEastWestRouteRG(name, ns, postfix string, r *eskip.Route) *eskip.Rout
+// of the internal domains and their CIDRs.
+func ParseEastWestRange(v []string) (map[string]string, error) {
+	eastWestRange := map[string]string{}
+	for _, i := range v {
```
i seems to be a bad name. Maybe cidr, but i is normally the index var not a value which is not an int

jonathanbeber replied:
I was operating with the index in the very beggining and forgot to update. Thanks!

szuecs commented on dataclients/kubernetes/eastwest.go:
```diff
@@ -43,3 +46,52 @@ func createEastWestRouteRG(name, ns, postfix string, r *eskip.Route) *eskip.Rout
+			return nil, fmt.Errorf("invalid east-west-range config: %s", i)
+		}
+		d, n := s[0], s[1]
+		_, net, err := net.ParseCIDR(n)
```
net shadows lib, cidr would be better.
d and n are as I commented above.

szuecs commented on dataclients/kubernetes/eastwest_test.go:
```diff
@@ -0,0 +1,180 @@
+			if e != nil {
+				if s.err != nil {
+					if e.Error() != s.err.Error() {
+						t.Error("invalid error received while parsing east-west range")
```
t.Errorf instead logs please

szuecs replied:
Same below multiple times

szuecs commented on dataclients/kubernetes/routegroup.go:
```diff
@@ -119,7 +120,10 @@ func toSymbol(p string) string {
-func rgRouteID(namespace, name, subName string, index, subIndex int) string {
+func rgRouteID(namespace, name, subName string, index, subIndex int, internal bool) string {
+	if internal {
+		name += "_internal"
```
Better prepend namespace

jonathanbeber replied:
thanks, I was not sure about that.

szuecs reviewed (changes requested):
Some nitpicks, but important ones.
<s>I think the tests should not all have the feature set, only the ones we add. </s>
The mergeHosts func deletion I would need to review more to judge good/bad, but as long tests pass we should be fine.
The configuration default should use nil instead of empty map.

jonathanbeber commented:
Thank you very much, @szuecs. I addressed the points but two:

> I think the tests should not all have the feature set, only the ones we add.

I tried to add tests for the scenarios the new code impacts, also the ones I believed should be checked with this new scenario of multiple internal and multiple external domains. I tried to base myself on the current east-west tests cases too.
Also, I believe I probably created scenarios that are not completed required. Maybe you could, please, point me the ones you believe are not required and we could discuss them? WDYT?

> The mergeHosts func deletion I would need to review more to judge good/bad, but as long tests pass we should be fine.

It was used only during route groups conversion for the catch-all routes that were created later. Since now it creates the catch-all 
right after the `transformRouteGroup` I used the `ctx.hostRoutes` directly instead of merging it. As you said, tests are passing and we should be fine.

Regarding the wrong predicate, I left a comment on the specific line for a clarification of your comment.

szuecs commented:
Adding tests are fine, but there are also modified tests (results modified), which means we don’t test the scenario without the eastwest range feature. So I would say duplicate some tests and add the eastwest range tests.

szuecs commented on config/config.go:
```diff
@@ -688,13 +696,19 @@ func (c *Config) Parse() error {
 
+	kubernetesEastWestRangePredicates, err := eskip.ParsePredicates(c.KubernetesEastWestRangePredicatesString)
+	if err != nil {
+		return fmt.Errorf("invalid east-west-range-predicates: %s", err.Error())
```
Better use %w if you create a wrapped error:
```
fmt.Errorf("invalid east-west-range-predicates: %w", err)
```

jonathanbeber commented:
:+1:

mikkeloscar commented:
:+1:

szuecs commented on dataclients/kubernetes/ingress.go:
```diff
@@ -47,6 +47,8 @@ type ingress struct {
 	kubernetesEnableEastWest    bool
 	eastWestDomainRegexpPostfix string
+	eastWestRangeDomains        []string
+	EastWestRangePredicates     []*eskip.Predicate
```
do we need to export this member?

jonathanbeber replied:
good catch

szuecs commented on dataclients/kubernetes/routegroup.go:
```diff
@@ -513,49 +518,120 @@ func transformRouteGroup(ctx *routeGroupContext) ([]*eskip.Route, error) {
 }
 
+func splitHosts(hosts []string, domains []string) (map[string]string, []string) {
+	i := map[string]string{}
```
i is more an index var, so I would suggest `h` instead

jonathanbeber replied:
it's for internal and shouldn't be a map since we changed to receive the predicates. Will fix it.

szuecs commented on dataclients/kubernetes/routegroup.go:
```diff
@@ -513,49 +518,120 @@ func transformRouteGroup(ctx *routeGroupContext) ([]*eskip.Route, error) {
 
+func splitHosts(hosts []string, domains []string) (map[string]string, []string) {
+	i := map[string]string{}
+	e := []string{}
```
not sure about `e`, maybe `a` ?

jonathanbeber replied:
it's it from internal and e for external. Will fix it.

jonathanbeber commented:
:+1:

szuecs commented on docs/tutorials/operations.md:
```diff
@@ -108,6 +108,73 @@ domain
+Alternatively, you can use Kubernetes East West Range feature. Use the
+flag `-kubernetes-east-west-range-domains` to define the cluster
+internal domains `-kubernetes-east-west-range-predicates` to define the
+[predicates](../../reference/predicates) that will be appended to every
```
../../reference/predicates -> ../../reference/predicates.md

szuecs commented on docs/tutorials/operations.md:
```diff
@@ -108,6 +108,73 @@ domain
 [known bug](https://github.com/zalando/skipper/issues/1024), if you
 combine it with custom routes.
 
+Alternatively, you can use Kubernetes East West Range feature. Use the
```
Maybe we should add a subsection for all the "more features"

jonathanbeber replied:
great idea, done!

szuecs commented:
:+1:

jonathanbeber commented:
:+1:

AlexanderYastrebov commented on dataclients/kubernetes/routegroup.go:
```diff
@@ -513,49 +518,120 @@ func transformRouteGroup(ctx *routeGroupContext) ([]*eskip.Route, error) {
+				hosts:                 externalHosts,
+				hostRx:                createHostRx(externalHosts...),
+				hostRoutes:            make(map[string][]*eskip.Route),
+				hasEastWestHost:       hasEastWestHost(r.options.KubernetesEastWestDomain, externalHosts),
```
I think this `hasEastWestHost` uses deprecated option `KubernetesEastWestDomain` 
https://github.com/zalando/skipper/blob/62f282f0f25a7aa50635e2b1f6956c6000cfa35b/dataclients/kubernetes/kube.go#L152-L154
and maybe not needed here at all  (i.e. always false) as hosts were split (`splitHosts`) before and we know `externalHosts` do not contains east-west hosts...

jonathanbeber replied:
As far as i understand we just deprecated but didn't remove it yet.

jonathanbeber replied:
Also, I deprecated it after merging this PR, so we might want to revisit it soon.

AlexanderYastrebov commented on dataclients/kubernetes/routegroup.go:
```diff
@@ -513,49 +518,120 @@ func transformRouteGroup(ctx *routeGroupContext) ([]*eskip.Route, error) {
-		mergeHostRoutes(hostRoutes, ctx.hostRoutes)
-	}
+		// Internal hosts
+		for _, host := range internalHosts {
```
What is the reason here to loop over internal domains where for external domains we just combine them
```go
hostRx:                createHostRx(externalHosts...),
```
@jonathanbeber 
Why not to do the same here, i.e. get rid of the loop and use single `internalCtx`
```go
hostRx:                createHostRx(internalHosts...),
```

jonathanbeber replied:
You're right, that was due to an [initial implementation](https://github.com/zalando/skipper/commit/b72f8d47cbda834395913171da2643dd466d0631#diff-3a9a80b20ebd6d5f62df3e5a8a080e2e022bb1ab9ec7a2089021f236f0f85b14R630) where we could define multiple IPs for different domains. The idea is that you could have `domain.internal` to the CIDR `10.0.0/8`, and `domain2.internal`  to the CIDR `192,168,0.0/24` for example. 

I will fix it.

AlexanderYastrebov replied:
Thank you

jonathanbeber replied:
#1758 should handle that.
